EXTENSIONS.md

```diff
@@ -0,0 +1,113 @@
+# Linkerd Extensions
+
+Linkerd has an extension model which allows 3rd parties to add functionality.
+Each extension consists of a binary CLI named `linkerd-name` (where `name` is
+the name of the extension) as well as a set of Kubernetes resources.  To invoke
+an extension, users can run `linkerd name` which will search the current PATH
+for an executable named `linkerd-name` and execute it.
+
+## Installing Extensions
+
+To install an extension, first download the extension executable and put it
+in your PATH. The extension can then be installed into your Kubernetes cluster
+by running `linkerd <extension name> install | kubectl apply -f -`. Similarly,
+it can be uninstalled by running
+`linkerd <extension name> uninstall | kubectl delete -f -`.
+
+A full list of installed extensions can be printed by running `linkerd check`.
+
+## Developing Extensions
+
+The extension must be an executable file named `linkerd-name` where `name` is
+the name of the extension. The name must not be any of the built-in Linkerd
+commands (e.g. `check`) or extensions (e.g. `viz`).
+
+The extension must accept the following flags and respect them any time that
+it communicates with the Kubernetes API.  All of these flags must be accepted
+but may be ignored if they are not applicable.
+
+* `--api-addr`: Override kubeconfig and communicate directly with the control plane at host:port (mostly for testing)
+* `--context`: Name of the kubeconfig context to use
+* `--help`/`-h`: Print help message
+* `--kubeconfig`: Path to the kubeconfig file to use for CLI requests
+* `--linkerd-namespace`/`-l`: Namespace in which Linkerd is installed [$LINKERD_NAMESPACE]
+* `--verbose`: Turn on debug logging 
+
+The extension must implement these commands:
```

Comment:
similar to the flags, i think these subcommands should be optional. they imply a specific type of extension that installs resources on a cluster. looking over kubectl's plugins, many (a majority?) seem to be cli-only: https://krew.sigs.k8s.io/plugins/

as a bonus, i think stateless / cli-only extensions create a lower barrier-to-entry for developers. i'd love to see an extension that reads linkerd metrics and outputs my network topology in graphviz format.

Reply:
That's fair.  This suggests two distinct types of extensions: installable extensions for which install, uninstall, and check must be implemented; and stateless extensions which operate without creating any resources in the cluster.